[AArch64][MachineCombiner] Reassociate long chains of accumulation instructions into a tree to increase ILP

jcohen-apple opened the pull request:
This pattern shows up often in media codes. The optimization should only kick in for O3. Currently only supports a single accumulation instruction, but can easily be expanded to support additional instructions in the future.

wangpc-pp commented:
I'm not familiar with AArch64 target so I added some reviewers and changed the title to increase visibility.

jcohen-apple commented:
Thanks, github suggested you and I wasn't sure who to add.

davemgreen reviewed (commented):
UABAL is a more unusual choice to pick as the first instruction. Did you try other operations like MLA, that are likely more common and I believe still perform the same accumulation? The idea sounds like a nice one, but adding extra instructions does sometimes make transforms like this less reliable than if we were purely reassociating.

Could the main part of the logic be added to base MachineCombiner, like the existing code that uses isAssociativeAndCommutative?

It is probably worth making sure there are tests for all the types, preferably with tests at the .ll level so we can see the whole output. Things like different chain lengths too, maybe, and signed/unsigned, loops and the "2" variants of the instructions.

jcohen-apple commented:
MLA is indeed more common, but less likely to get linearized by the compiler. See this godbolt example where the compiler linearizes multiple accumulators due to reassociate pass running after inlining a function - https://godbolt.org/z/1aPbb3YY1. This same bug should appear for other intrinsics whose ISEL pattern also contains `add` instructions. If you find it confusing that the first commit handles UABAL and not something more obvious, I can start with MLA and then add UABAL.
Adding the additional instruction does indeed make it slightly more tricky than the other patterns handled by MachineCombiner. I thought that the checks on profitability of the pattern (that it reduces critical path length and does not increase resource length) still apply here. 
I'll move the main logic into MachineCombiner pass and add the tests you suggested, and MLA as well.

davemgreen commented:
I was mostly worried about UABAL not being generated very often, and this turning out to not be profitable because we hadn't tested it on a larger number of cases. With how many ops are needed to make it profitable (8?) it seems like it should be OK. It is probably worth adding SABAL too though, to keep them symmetric.

I wasn't sure how the end of the chain was calculated, I thought it might be worth looking at a looping chain too as in https://godbolt.org/z/8conT4xaa. It might need a longer chain to be profitable though, longer than might be realistic if it loses the ability to forward the uabal (if that is how it works https://godbolt.org/z/YWhY98dxj). Perhaps it needs a higher-level transform that could accumulate multiple chains after the loop.

And it might be more awkward, but is it worth trying to support uabal and uabal2 operations in the same chain? I imagine the interleaving of the high instructions to be quite efficient in well written vector code that can use them.

jcohen-apple commented:
@davemgreen Made most of the changes you suggested. For this pattern to apply on MLA instructions I had to modify machine-combiner to run recursively while there are still changes to apply to the basic block. Let me know what you think of this approach before I expand the pattern to additional data types.

davemgreen commented on llvm/lib/CodeGen/TargetInstrInfo.cpp:
```diff
@@ -899,6 +913,152 @@ bool TargetInstrInfo::isReassociationCandidate(const MachineInstr &Inst,
+
+  while (CurrentInstr &&
+         (canCombine(MBB, CurrentInstr->getOperand(1), AccumulatorOpcode) ||
+          canCombine(MBB, CurrentInstr->getOperand(1),
```
This looks like it does duplicate work, but could just check the opcode for the two values?

davemgreen commented on llvm/lib/CodeGen/TargetInstrInfo.cpp:
```diff
@@ -899,6 +913,152 @@ bool TargetInstrInfo::isReassociationCandidate(const MachineInstr &Inst,
+  }
+
+  // Add the instruction at the top of the chain.
+  if (CurrentInstr->getOpcode() == ChainStartOpCode.value())
```
What happens if we get to the start of a chain and it isn't a ChainStartOpCode?

jcohen-apple replied:
Good catch. I was being too specific when searching for this pattern. Updated with a relevant test case and modified some of the code to support this.

davemgreen commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -6674,6 +6674,118 @@ static bool getMaddPatterns(MachineInstr &Root,
+    unsigned AccumulationOpcode) const {
+  switch (AccumulationOpcode) {
+  default:
+    llvm_unreachable("Unknown accumulator opcode");
```
If these do llvm_unreachable's, do they need to return an optional?

jcohen-apple replied:
No, leftovers from an earlier version of the code. Removed. Thanks!

davemgreen replied:
I was actually thinking the other way - to remove optional and make it an error to call these without isAccumulationOpcode returning true. IIUC it might start to go wrong it they returned nullopt when we got this far, or they assert already.  (Maybe isAccumulationOpcode could return `optional<pair<AccumulationStartOpcode, ReduceOpcodeForAccumulator>>` so that all the info is together in one place, but I'm not sure that would be better considering ReduceOpcodeForAccumulator can return the same value for opcodes).

davemgreen reviewed (commented):
> For this pattern to apply on MLA instructions I had to modify machine-combiner to run recursively while there are still changes to apply to the basic block. Let me know what you think of this approach before I expand the pattern to additional data types.

I gave it a test (with udot too) and I think it looked OK. Thanks for the updates.

The recursion is because there are multiple MLA patterns? Often routines like instcombine have a worklist that they can re-add to if needed. Perhaps so long as we know this it's causing problems lets get the UABA and SABA in first then, and we can add MLA with a later patch if needed where we can sort out re-visiting nodes.

davemgreen commented on llvm/lib/CodeGen/TargetInstrInfo.cpp:
```diff
@@ -1252,19 +1425,101 @@ void TargetInstrInfo::genAlternativeCodeSequence(
+                  .addReg(Instr->getOperand(3).getReg(),
+                          getKillRegState(Instr->getOperand(3).isKill()));
+      } else {
+        // For the remaining cases, we need ot use an output register of one of
```
ot -> to

davemgreen commented on llvm/lib/CodeGen/TargetInstrInfo.cpp:
```diff
@@ -899,6 +913,158 @@ bool TargetInstrInfo::isReassociationCandidate(const MachineInstr &Inst,
+  std::optional<unsigned> ReduceOpCode =
+      getReduceOpcodeForAccumulator(Root.getOpcode());
+
+  if (!ReduceOpCode.value())
```
For example, I believe if we get here we have already made alterations to the code, so it might be bad if we decide to bail out.

jcohen-apple replied:
I assumed that the code alteration will only be applied if the code pattern is applied successfully, but I agree it makes more sense to bail if this is used incorrectly. I changed them back to use only `llvm_unreachable`

davemgreen commented on llvm/lib/CodeGen/TargetInstrInfo.cpp:
```diff
@@ -899,6 +913,158 @@ bool TargetInstrInfo::isReassociationCandidate(const MachineInstr &Inst,
+    InsInstrs.push_back(MIB);
+  }
+
+  // If the number of registers to reduce is odd, add the reminaing register to
```
remaining

davemgreen commented on llvm/lib/CodeGen/TargetInstrInfo.cpp:
```diff
@@ -899,6 +913,158 @@ bool TargetInstrInfo::isReassociationCandidate(const MachineInstr &Inst,
+// A chain of accumulation instructions will be selected IFF:
+//    1. All the accumulation instructions in the chain have the same opcode,
+//       besides the first that has a slightly different opcode because it does
+//       not perform the accumulation, just defines it.
```
defines it -> begins it?

jcohen-apple committed:
[Machine-Combiner] Add a pass to reassociate chains of accumulation instructions into a tree

This pass is designed to increase ILP by performing accumulation into multiple registers. It currently supports only the S/UABAL accumulation instruction, but can be extended to support additional instructions.

jcohen-apple commented on llvm/test/CodeGen/AArch64/aarch64-reassociate-accumulators.ll:
```diff
@@ -0,0 +1,530 @@
+  ret i32 %reduce
+}
+
+; CHECK-LABEL: uabal_i16_to_i32_accumulation
```
Done

jcohen-apple commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -6674,6 +6675,115 @@ static bool getMaddPatterns(MachineInstr &Root,
+  switch (Opcode) {
+  default:
+    break;
+  case AArch64::UABALB_ZZZ_D:
```
Done

davemgreen reviewed (approved):
Thanks - LGTM
